Wrappers/Python/cil/optimisation/algorithms/Algorithm.py

```diff
@@ -16,13 +16,69 @@
 #
 # Authors:
 # CIL Developers, listed at: https://github.com/TomographicImaging/CIL/blob/master/NOTICE.txt
+from abc import ABC, abstractmethod
+from numbers import Integral
+from typing import Optional
+from warnings import warn
 
-import time, functools
-from numbers import Integral, Number
-import logging
 import numpy as np
+from tqdm.auto import tqdm as tqdm_auto
 
-class Algorithm(object):
+
+class Callback(ABC):
+    '''Base Callback to inherit from for use in :code:`Algorithm.run(callbacks: list[Callback])`.'''
+    def __init__(self, verbose=1):
+        ''':param verbose: int, 0=quiet, 1=info, 2=debug'''
+        self.verbose = verbose
+
+    @abstractmethod
+    def __call__(self, algorithm):
+        pass
+
+
+class OldCallback(Callback):
+    '''Converts an old-style :code:`def callback(iteration, objective, x)`
+    to a new-style :code:`class Callback`.
+    '''
+    def __init__(self, callback, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.func = callback
+
+    def __call__(self, algorithm):
+        if algorithm.update_objective_interval > 0 and algorithm.iteration % algorithm.update_objective_interval == 0:
+            self.func(algorithm.iteration, algorithm.get_last_objective(return_all=self.verbose), algorithm.x)
+
+
+class ProgressCallback(Callback):
+    ''':code:`tqdm`-based progress bar.'''
+    def __init__(self, verbose=1, tqdm_class=tqdm_auto, **tqdm_kwargs):
+        '''
+        :param tqdm_kwargs: passed to :code:`tqdm_class`
+          (e.g. :code:`file=some_logfile`).
+        '''
+        super().__init__(verbose=verbose)
+        self.tqdm_class = tqdm_class
+        self.tqdm_kwargs = tqdm_kwargs
+
+    def __call__(self, algorithm):
+        if not hasattr(self, 'pbar'):
+            tqdm_kwargs = self.tqdm_kwargs
+            tqdm_kwargs.setdefault('total', algorithm.max_iteration)
+            tqdm_kwargs.setdefault('disable', not self.verbose)
+            self.pbar = self.tqdm_class(**tqdm_kwargs)
+        self.pbar.set_postfix(algorithm.objective_to_dict(self.verbose>=2), refresh=False)
+        self.pbar.update(algorithm.iteration - self.pbar.n)
+
+
+class LogfileCallback(ProgressCallback):
+    ''':code:`ProgressCallback` but to a file instead of screen'''
+    def __init__(self, log_file=None, mode='a', mininterval=5, **kwargs):
+        """:param miniterval: approximate number of seconds between updates."""
```

Comment:
seconds?

Reply:
yes. if there are ~20 iterations per second it'll only print every ~100 iterations.

Reply:
That's interesting! We'll never have that!

Reply:
if there are 0.4 iterations per second, then "5 seconds" means it'll only print every ~2 iterations :)